weaver/wps_restapi/jobs/utils.py

```diff
@@ -496,69 +631,445 @@ def get_job_results_response(job, container, headers=None):
+    if out_mode == ExecuteTransmissionMode.REFERENCE:
+        link = make_result_link(job, result, output_id, out_mode, output_format, settings=settings)
+        headers.extend([("Link", link[0])])
+        return HTTPNoContent(headers=headers)
+
+    # convert value as needed since reference transmission was not requested/resolved
+    out_headers = {}
+    if is_ref:
+        output_mode = ExecuteTransmissionMode.VALUE
+        out_headers, out_data = generate_or_resolve_result(
+            job,
+            result,
+            output_id,
+            output_id,
+            output_mode,
+            output_format,
+            settings=settings,
+        )
+        headers.update(out_headers)
+
+    ctype = out_headers.get("Content-Type")
+    if not ctype:
+        ctype = get_field(result, "mediaType", search_variations=True, default=ContentType.TEXT_PLAIN)
+    c_enc = cast("AnyContentEncoding", headers.get("Content-Encoding") or "UTF-8")  # type: AnyContentEncoding
+    out_data = data2str(out_data)
+    out_data = ContentEncoding.encode(out_data, c_enc)
+    return HTTPOk(body=out_data, content_type=ctype, charset=c_enc, headers=headers)
+
+
+def get_job_results_document(job, results, *, settings):
+    # type: (Job, ExecutionResults, Any, SettingsType) -> ExecutionResults
+    """
+    Generates the :term:`Job` results document response from available or requested outputs with necessary conversions.
+
+    Removes nested literal value definitions if qualified value representation is not needed.
+    Qualified value representation is not needed if no other field than ``value`` is provided with the literal data,
+    or when the specified :term:`Media-Type` is simply the plain text default for data literals.
+    The simplification is applied for both literals on their own and nested array of literals.
+    However, when processing an array, the qualified value representation is preserved if any of the items requires
+    the explicit mention of another :term:`Media-Type` than plain text, to return a consistent structure.
+
+    Uses the :paramref:`job` definition and submitted ``headers``
+
+    .. warning::
+        This function assumes that schema deserialization was applied beforehand.
+        Therefore, it will not attempt matching every possible combination of the results representation.
+    """
+
+    def make_result(result, result_id, output_id):
+        # type: (ExecutionResultValue, str, str) -> Union[AnyValueType, ExecutionResultObject]
+        if isinstance(result, dict):
+            is_ref = bool(get_any_value(result, key=True, file=True, data=False))
+            val = get_any_value(result)
+        else:
+            is_ref = False
+            val = result
+            result = {"value": val}
+        out_mode, out_fmt = get_job_output_transmission(job, result_id, is_reference=is_ref)
+        headers, data = generate_or_resolve_result(job, result, result_id, output_id, out_mode, out_fmt, settings)
```

Comment:
Why not use directly `get_job_results_single`

Reply:
They do similar operations, but `get_job_results_single` could potentially do other things not needed by the document representation if more capabilities are added alter on.